types/schema.d.ts

```diff
@@ -14,4 +14,4 @@ export interface FastifySchema {
 /**
  * Compiler for FastifySchema Type
  */
-export type FastifySchemaCompiler = (schema: FastifySchema) => unknown
+export type FastifySchemaCompiler = (method: string, url: string, httpPart: string, schema: FastifySchema) => unknown
```

Comment:
@Ethan-Arrowood I'm not sure, but the `export interface FastifySchema` could be wrong cause the user will receive as input a plain schema, not the complete schema configuration

```
fastify.post('/the/url', {
  body: bodyJsonSchema,
  querystring: queryStringJsonSchema,
  params: paramsJsonSchema,
  headers: headersJsonSchema
}, handler)
```

The user will get 4 calls with `bodyJsonSchema, queryStringJsonSchema, paramsJsonSchema, headersJsonSchema` each one

Could you help me with the TS test? I need only some suggestions 😇

Reply:
I'm confused -- i thought those properties fall under the `schema` property on the route options object. I noticed in your unit tests theres an `reponse` property now added to the schema so that should be added to the `FastifySchema` interface. 

These schema properties are specified as type `unknown` because this allows the user to tell TypeScript what they are _strictly_. If we used something like `{ [k: string]: any }` then we are making the type system non-strict by default and that does not lead to a great user experience. Check out the [Schema](https://github.com/Ethan-Arrowood/fastify/blob/typescript-docs/docs/TypeScript.md#json-schema) section in my TS docs for more info.

Reply:
>  i thought those properties fall under the schema property on the route options object

This is correct

> in your unit tests theres an reponse property now added to the schema so that should be added to the FastifySchema interface.

This feature is already in place [from master](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#serialization)

My concern is that the `FastifySchemaCompiler` want a `FastifySchema` as input (`schema` parameter) but this parameter is a JSON Schema.
I think we need a different type here since `FastifySchema` works for route's configuration only.


I'll try to explain me by example:

```
// want a JSON schema input
  fastify.addSchema({
    $id: 'test',
    type: 'object',
    properties: {
      id: { type: 'number' }
    }
  })

  // want a `FastifySchema` type
  fastify.get('/:id', {
    handler: (req, reply) => {
      reply.send({ id: req.params.id * 2, ignore: 'it' })
    },
    schema: {
      params: { $ref: 'test#' },
      response: {
        200: { $ref: 'test#' }
      }
    }
  })

fastify.setValidatorCompiler((method, url, httpPart, schema) => {
 // looking the route above, the args are:
- httpPart = "params"
- schema = { $ref: 'test#' }
})
```

Reply:
this looks good to me now